### PR TITLE
Fix `QNSPSA` Optimizer

### LINKED ISSUE #5437: [BUG] `QNSPSAOptimizer` with Autograd changes differentiable parameters type to vanilla NumPy array
### Expected behavior

Updating parameters with `opt = qml.QNSPSAOptimizer` should not change the type of the parameters.

### Actual behavior

The parameters change from `<class 'pennylane.numpy.tensor.tensor'>` to `<class 'numpy.ndarray'>` after an optimization step.

### Additional information

Stems from this forum post: https://discuss.pennylane.ai/t/qnspsa-error-using-a-dataset/4153/12
Related PR: https://github.com/PennyLaneAI/pennylane/pull/4920

### Source code

```shell
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit")
@qml.qnode(dev)
def cost(params):
    qml.RX(params[0], wires=0)
    qml.CRY(params[1], wires=[0, 1])
    return qml.expval(qml.Z(0) @ qml.Z(1))

params = np.random.rand(2)
opt = qml.QNSPSAOptimizer(stepsize=5e-2)

for _ in range(3):
    print(type(params))
    params, loss = opt.step_and_cost(cost, params)
    #params = np.array(params) # this fixes the error
```

### Tracebacks

```shell
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[17], line 16
     14 for _ in range(3):
     15     print(type(params))
---> 16     params, loss = opt.step_and_cost(cost, params)
     17     #params = np.array(params) # this fixes the error

File ~/.virtualenvs/pennylane-stable/lib/python3.11/site-packages/pennylane/optimize/qnspsa.py:185, in QNSPSAOptimizer.step_and_cost(self, cost, *args, **kwargs)
    172 def step_and_cost(self, cost, *args, **kwargs):
    173     r"""Update trainable parameters with one step of the optimizer and return
    174     the corresponding objective function value after the step.
    175 
   (...)
    183         function output prior to the step
    184     """
--> 185     params_next = self._step_core(cost, args, kwargs)
    187     if not self.blocking:
    188         loss_curr = cost(*args, **kwargs)

File ~/.virtualenvs/pennylane-stable/lib/python3.11/site-packages/pennylane/optimize/qnspsa.py:215, in QNSPSAOptimizer._step_core(self, cost, args, kwargs)
    213 grad_tapes, grad_dirs = self._get_spsa_grad_tapes(cost, args, kwargs)
    214 # metric_tapes contains 4 tapes for tensor estimation
--> 215 metric_tapes, tensor_dirs = self._get_tensor_tapes(cost, args, kwargs)
    216 all_grad_tapes += grad_tapes
    217 all_metric_tapes += metric_tapes

File ~/.virtualenvs/pennylane-stable/lib/python3.11/site-packages/pennylane/optimize/qnspsa.py:408, in QNSPSAOptimizer._get_tensor_tapes(self, cost, args, kwargs)
    406     args_list[2][index] = arg + self.finite_diff_step * (-dir1 + dir2)
    407     args_list[3][index] = arg - self.finite_diff_step * dir1
--> 408 dir_vecs = (np.concatenate(dir1_list), np.concatenate(dir2_list))
    409 tapes = [
    410     self._get_overlap_tape(cost, args, args_finite_diff, kwargs)
    411     for args_finite_diff in args_list
    412 ]
    414 return tapes, dir_vecs

ValueError: need at least one array to concatenate
```

### System information

```shell
Name: PennyLane
Version: 0.35.1
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/isaac/.virtualenvs/pennylane-stable/lib/python3.11/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane_Lightning

Platform info:           macOS-14.4-arm64-arm-64bit
Python version:          3.11.8
Numpy version:           1.26.4
Scipy version:           1.12.0
Installed devices:
- default.clifford (PennyLane-0.35.1)
- default.gaussian (PennyLane-0.35.1)
- default.mixed (PennyLane-0.35.1)
- default.qubit (PennyLane-0.35.1)
- default.qubit.autograd (PennyLane-0.35.1)
- default.qubit.jax (PennyLane-0.35.1)
- default.qubit.legacy (PennyLane-0.35.1)
- default.qubit.tf (PennyLane-0.35.1)
- default.qubit.torch (PennyLane-0.35.1)
- default.qutrit (PennyLane-0.35.1)
- null.qubit (PennyLane-0.35.1)
- lightning.qubit (PennyLane_Lightning-0.35.1)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes #5437 .  [sc-59838]

When we started distinguishing vanilla numpy and autograd numpy in our source code, we accidentally switched to using vanilla numpy in the `QNSPSA` optimizer instead of autograd numpy.  This switches it back.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -291,6 +291,8 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* The `QNSPSAOptimizer` can now be used for more than one step.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:294]:
```suggestion
* The `QNSPSAOptimizer` can now be used for more than one step.
  [(#5439)](https://github.com/PennyLaneAI/pennylane/pull/5439)
```

**Review by isaacdevlugt — APPROVED** [review]
Thanks Christina! The fix works for me.

**Review by PietropaoloFrisoni — APPROVED** [review]
Thanks!

**Review by astralcai — COMMENTED** [review]
I have a question: I see that here `pennylane.numpy` is used in all cases instead of vanilla `numpy`. Is this fine or should we differentiate based on whether trainability is needed for each particular case?

**Comment by albi3ro** [comment]
> I have a question: I see that here `pennylane.numpy` is used in all cases instead of vanilla `numpy`. Is this fine or should we differentiate based on whether trainability is needed for each particular case?

@astralcai  Something we could potentially look into. Some of the variables we are creating might not actually end up being trainable, and could just being intermediaries.  But the optimizers are autograd-only, so we don't have to worry about mixing ml-framework types. I'm not as concerned about using autograd numpy as I would be in other parts of the code base.

**Pull request merged**